source/common/network/dns_impl.cc

```diff
@@ -84,6 +84,13 @@ void DnsResolverImpl::PendingResolution::onAresGetAddrInfoCallback(int status, i
   // We receive ARES_EDESTRUCTION when destructing with pending queries.
   if (status == ARES_EDESTRUCTION) {
     ASSERT(owned_);
+    // This destruction might have been triggered by a peer PendingResolution that received a
+    // ARES_ECONNREFUSED. If the PendingResolution has not been cancelled that means that the
+    // callback_ target _should_ still be around. In that case, raise the callback_ so the target
+    // can be done with this ActiveQuery and initiate a new one.
+    if (!cancelled_) {
```

Comment:
This approach relies on the fact that all callback targets (strict_dns_cluster, logical_dns_cluster, redis_cluster, dns_cache) cancels active queries in their destructors, which is already used when a PendingResolution successfully completes. However, this is not an enforced contract.

I was wondering if we wanted to instead go further and move this to a similar pattern as the one in the init manager where the lifetimes of the target and the watcher are independent of each other and it is safe to call functions/callbacks on each other regardless of them being freed.

Reply:
I think this is fine but if you want to add a TODO to clean this up, up to you.

Reply:
👌 #10079